### PR TITLE
build: INFENG-780: Remove CircleCI bumpversion

### PR DESCRIPTION
**Pull request by davidfluck-hpe**
## Ticket
INFENG-780

## Description
This changeset does a few things, and I apologize that it's all over the place:

First, I removed the `bump2version` ("bumpversion", colloquially) dependency, both literally from `requirements.txt` and also the last remaining reference in its config file, `.bumpversion.cfg`, along with the config file itself. The remaining reference was in `.circleci/real_config.yml`. Removing it, though, was somewhat tricky, and had a few knock-on effects that had to be solved.

The version string itself was a default for the pipeline parameter `det-version`, which is threaded through the entirety of the build process as a source of truth for the Determined version information. Instead of updating the CircleCI config itself, we can dynamically generate it in `.circleci/config.yml` and set the parameter there. There's still some work left to do to get tag builds working correctly, but for now, I call `version.sh` to set `det-version`, which should work correctly for all branch builds, by returning the local version string, e.g. `0.34.0+2e616a3d7`

I had to tweak `version.sh` to fix a small bug that caused git tags with subsequent `v`s in them, like `v1.0.3-dev0` to be munged incorrectly into `1.0.3-de0`. Now, I invoke `sed` to remove only a single `v` at the beginning of the string.

I updated two direct invocations of `setup.py` in CircleCI to instead use the `build` module, as invoking `setup.py` directly is deprecated.

I updated the agent creation code (I believe) to URL-escape the query parameter that's used to set the agent version. Under the new versioning scheme, local build versions use a short SHA after a plus `+` for uniqueness, and the plus sign was being interpreted as being part of a URL, where it was then turned into a space. This broke tests that attempt to fetch the agent version (it's also just plain incorrect). Now, the version string is properly escaped, and the proper agent version is returned.

I had to overhaul how `setup.py` and the `determined` (`harness/`) Python package manage versions. I kept running into test errors that indicated that the correct version wasn't being set somewhere in setuptools. At first, I mistakenly assumed that `pytest` was, somehow, using the bare source for testing. This is incorrect, though (and makes no sense anyway). `pytest` still needs to import packages, and it does so after we `pip install` an editable install of `determined`. Editable installs still end up calling setuptools, which means the built package has package metadata, and so we can still use `importlib.metadata` locally to fetch version information.

The only bit that really had to be changed was `harness/setup.py` and `model_hub/setup.py`. I did tweak `harness/determined/__init__.py`, but that's just to catch the possible `importlib.metadata.PackageNotFoundError` exception. I believe this will never happen, but I've been wrong before. And there really is no excuse for `__init__.py` to ever do anything that isn't successful. So now, and it feels mildly gross, but is not too different from how other large projects like `numpy` manage versions, `setup.py` simply shells out to `version.sh` to get a version number during package build, if `VERSION` isn't already set. I realize that I also check `VERSION` in `version.sh`, but a belt _and_ suspenders won't hurt. If `VERSION` is set, we use that; otherwise, we get the local (`+SHA`) version.

This enables us to do editable installations, which a bunch of things assume anyway, without having to ensure `VERSION` is set. Before, `setup.py` assumed `VERSION` was set, or else it would error out. But when you `pip install -e harness` directly, like in CI or even locally (which can happen with `pip install -r requirements.txt`, as `-e harness` is the first line), that variable won't necessarily be set. So instead, `setup.py` checks `VERSION`, and then, if it's empty, attempts to fetch the version itself.

If `setup.py` and `__init__.py` can't retrieve versions correctly, they'll return default values. `setup.py` will return `1!0.0.0+unknown`, whereas `__init__.py` will return `2!0.0.0+unknown`. My hope is that using two different versions can help indicate where the error is, and using epochs should make package requirement resolution still pass, as both `1` and `2` are greater than the implicit `0` epoch that our version numbers use now.

I did some other administrivia, like removing or updating comments referencing `bumpversion`, adding an additional `VERSION` environment variable to a part of the CircleCI config, running `black` on a file it was complaining about which has since been fixed upstream, and added return types to the functions in `setup.py` to satisfy `mypy`.

## Test Plan
As always, this one is kind of tricky. Most of this work involved getting the version logic actually correct, and then fixing broken tests/other random broken things, so this might be more of a spot check. If you go to CircleCI, though, and find the last CI run for this branch, it should be all green. I would say try a local build as well, both with and without `VERSION` set to an explicit version string (which doesn't have to begin with `v`, but the script can handle both). You can also tag either `HEAD` or a few commits back, then run `version.sh` and note the value to compare later.

Once `determined` is built and installed locally, open something like `ipython` and attempt to import determined, then try `determined.__version__`. It should match the version in `VERSION`, or the version returned by `version.sh`.

Basically, try to exercise any of the normal paths you would when working with the Determined codebase, and make sure they still work, and that version numbers still match and are sane.

## Checklist

- [ ] Changes have been manually QA'd
- [ ] New features have been approved by the corresponding PM
- [ ] User-facing API changes have the "User-facing API Change" label
- [ ] Release notes have been added as a separate file under `docs/release-notes/`
  See [Release Note](https://github.com/determined-ai/determined/blob/master/docs/release-notes/README.md) for details.
- [ ] Licenses have been included for new code which was copied and/or modified from any external code

### CODE REVIEW

**Review by tara-det-ai — APPROVED** [review]
LGTM

File: agent/internal/agent.go
Comment (dannysauer) [inline agent/internal/agent.go:1]: Are these changes intended to be part of this PR?  Just double-checking. :)

File: version.sh [hunk 1]
```diff
@@ -53,28 +53,29 @@ if [ -z ${VERSION} ]; then
...
+        )
     fi
 
     # Munge the tag into the form we want. Note: we always append a SHA hash,
     # even if we're on the commit with the tag. This is partially because I feel
     # like it will be more consistent and result in fewer surprises, but also it
     # might help indicate that this is a local version. Additionally, remove the
     # 'v' from the final version string.
-    echo -n "${MAYBE_TAG}+${SHA}" | tr -d 'v'
+    echo -n "${MAYBE_TAG}+${SHA}" | sed -e 's/^v//'
```
Comment (dannysauer) [inline version.sh:73]: This is bash, so I suppose I'll just throw this one out there too. ;)
```suggestion
    echo -n "${MAYBE_TAG#v}+${SHA}"
```
I do like using sed to make sure it's at the beginning rather than "anywhere at all" though.

File: version.sh [hunk 2]
```diff
@@ -53,28 +53,29 @@ if [ -z ${VERSION} ]; then
...
 else
     # Use existing VERSION, which is much easier. This should be the default
     # case for CI, as VERSION will already be set. We also remove the 'v' from
     # the tag for the version string, as that is what the current CI
     # functionality expects. Finally, use tr to remove the 'v' prefix to get the
     # bare version string.
-    echo -n "${VERSION}" | tr -d 'v'
+    echo -n "${VERSION}" | sed -e 's/^v//'
```
Comment (dannysauer) [inline version.sh:80]: See previous bash-native comment. https://github.com/determined-ai/determined/pull/9785/files#r1704204617
```suggestion
    echo -n "${VERSION#v}"
```

Also, note that the comment above this code says "finally, use tr..." which is inaccurate either way. :)

**Review by dannysauer — APPROVED** [review]
This looks fine to me, though verify that go file change is necessary, and it might be worthwhile to make the comment match the code even if the bash-native string manipulation is not used. :)

File: agent/internal/agent.go
Comment (davidfluck-hpe) [reply agent/internal/agent.go:1]: Yeah, that's for INFENG-786. It really should've been a separate PR, but I was on a warpath fixing tests, including the ones that were broken because of the agent version string being returned incorrectly.

File: version.sh (hunk 1 above)
Comment (davidfluck-hpe) [reply version.sh:73]: Huh, I learned a new shell expansion operator today! I checked some cases and it seems like it only matches at the front.
